float: Replace some approximate assertions with exact

tgross35 opened the pull request:
As was mentioned at [1], we currently use `assert_approx_eq` for testing
some math functions that guarantee exact results. Replace approximate
assertions with exact ones for the following:

* `ceil`
* `floor`
* `fract`
* `from_bits`
* `mul_add`
* `round_ties_even`
* `round`
* `trunc`

This likely wasn't done in the past to avoid writing out exact decimals
that don't match the intuitive answer (e.g. 1.3 - 1.0 = 0.300...004),
but ensuring our results are accurate seems more important here.

[1]: https://github.com/rust-lang/rust/pull/138087#issuecomment-2842069281

The first commit is a small bit of macro cleanup.

try-job: aarch64-gnu
try-job: x86_64-gnu-aux

<!-- homu-ignore:start -->
cc @beetrees as you suggested this
<!-- homu-ignore:end -->

tgross35 committed:
float: Use a shared `assert_biteq!` macro for tests

Clean up the separate `assert_f{16,32,64,128}` macros with a single
`assert_biteq!` macro that works for all float widths.

tgross35 committed:
float: Replace some approximate assertions with exact

As was mentioned at [1], we currently use `assert_approx_eq` for testing
some math functions that guarantee exact results. Replace approximate
assertions with exact ones for the following:

* `ceil`
* `floor`
* `fract`
* `from_bits`
* `mul_add`
* `round_ties_even`
* `round`
* `trunc`

This likely wasn't done in the past to avoid writing out exact decimals
that don't match the intuitive answer (e.g. 1.3 - 1.0 = 0.300...004),
but ensuring our results are accurate seems more important here.

[1]: https://github.com/rust-lang/rust/pull/138087#issuecomment-2842069281

tgross35 commented:
@bors try

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15289455298/job/43006167779) [(plain)](https://github.com/rust-lang/rust/commit/f12c89ea817b20b4bb95e17050b40245677f00cc/checks/43006167779/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 22.5s done
#18 DONE 30.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<f128, f128>
             at /checkout/library/core/src/panicking.rs:387:5
   4: coretests::floats::f128::test_fract
             at ./tests/floats/f128.rs:335:5
   5: coretests::floats::f128::test_fract::{closure#0}
             at ./tests/floats/f128.rs:333:16
---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<f128, f128>
             at /checkout/library/core/src/panicking.rs:387:5
   4: coretests::floats::f128::test_mul_add
             at ./tests/floats/f128.rs:454:5
   5: coretests::floats::f128::test_mul_add::{closure#0}
             at ./tests/floats/f128.rs:450:18
---
    floats::f128::test_mul_add

test result: FAILED. 1971 passed; 2 failed; 4 ignored; 0 measured; 0 filtered out; finished in 1.36s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.89.0-nightly\t(c3d1bd635\t2025-05-27)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "x86_64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:25:40
  local time: Wed May 28 02:07:14 UTC 2025
  network time: Wed, 28 May 2025 02:07:14 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15289471444/job/43006210702) [(plain)](https://github.com/rust-lang/rust/commit/6f9d015f59c1fed80edc537102cb12d93e11ffd6/checks/43006210702/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<f128, f128>
             at /checkout/library/core/src/panicking.rs:387:5
   4: coretests::floats::f128::test_fract
             at ./tests/floats/f128.rs:335:5
   5: coretests::floats::f128::test_fract::{closure#0}
             at ./tests/floats/f128.rs:333:16
---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<f128, f128>
             at /checkout/library/core/src/panicking.rs:387:5
   4: coretests::floats::f128::test_mul_add
             at ./tests/floats/f128.rs:454:5
   5: coretests::floats::f128::test_mul_add::{closure#0}
             at ./tests/floats/f128.rs:450:18
---
    floats::f128::test_mul_add

test result: FAILED. 1971 passed; 2 failed; 4 ignored; 0 measured; 0 filtered out; finished in 2.68s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.89.0-nightly\t(6f9d015f5\t2025-05-28)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/lib\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:57:00
  local time: Wed May 28 02:41:37 UTC 2025
  network time: Wed, 28 May 2025 02:41:37 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15289471444/job/43008257000)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/15289471444/job/43008257000","builder_name":"checks-actions"} -->

tgross35 commented:
@beetrees any idea what is going on here? It looks like the f128 fma tests aren't giving the expected results with clif.

tgross35 commented on library/coretests/tests/floats/f128.rs:
```diff
@@ -462,10 +451,10 @@ fn test_mul_add() {
-    assert_approx_eq!((-12.3f128).mul_add(-4.5, -6.7), 48.65, TOL_PRECISE);
-    assert_approx_eq!(0.0f128.mul_add(8.9, 1.2), 1.2, TOL_PRECISE);
-    assert_approx_eq!(3.4f128.mul_add(-0.0, 5.6), 5.6, TOL_PRECISE);
+    assert_eq!(12.3f128.mul_add(4.5, 6.7), 62.0500000000000033750779948604758829);
+    assert_eq!((-12.3f128).mul_add(-4.5, -6.7), 48.65000000000000301980662698042579);
```
🙃 of course, that's why my results from Julia weren't working either. Thanks for the fixes.

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit b94d32eab0c093978e5bf7e4f391a2b100560277 with merge f2c471b102998ac17e9717af5c6da4647473717b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b94d32eab0c093978e5bf7e4f391a2b100560277","merge_sha":"f2c471b102998ac17e9717af5c6da4647473717b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15310085858/job/43078516747)
Build commit: f2c471b102998ac17e9717af5c6da4647473717b (`f2c471b102998ac17e9717af5c6da4647473717b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15310085858/job/43078516747"},"merge_sha":"f2c471b102998ac17e9717af5c6da4647473717b"} -->

RalfJung commented on library/coretests/tests/floats/f128.rs:
```diff
@@ -260,98 +249,98 @@ fn test_classify() {
 #[cfg(not(miri))]
```
Miri has f16 and f128 implementations for all the rounding functions by now -- I don't think there's still a good reason for these tests to be disabled in Miri?

tgross35 replied:
Not at all, I just wasn't aware of the change. I added 2e4824d8b4a8b1a33908f7af702b07d8871de544 to remove the `#[cfg]`.

(which try job runs miri again?)

RalfJung replied:
> (which try job runs miri again?)

The gnu-aux one.

tgross35 committed:
float: Enable some f16 and f128 rounding tests on miri

The rounding tests are now supported, so there is no longer any reason
to skip these.

tgross35 committed:
float: Use `assert_biteq!` where possible

`assert_eq!` ignores the sign of zero, but for any tests involving zeros
we do care about this sign. Replace `assert_eq!` with `assert_biteq!`
everywhere possible for float tests to ensure we don't miss this.
`assert_biteq!` is also updated to check equality on non-NaNs, to catch
the unlikely case that bitwise equality works but our `==`
implementation is broken.

There is one notable output change: we were asserting that
`(-0.0).fract()` and `(-1.0).fract()` both return -0.0, but both
actually return +0.0.

RalfJung commented on library/coretests/tests/floats/f128.rs:
```diff
@@ -257,114 +247,107 @@ fn test_classify() {
 
 #[test]
-#[cfg(not(miri))]
 #[cfg(target_has_reliable_f128_math)]
```
Ideally this would become `any(miri, target_has_reliable_f128_math)` since f16/f128 math is always reliable with Miri... not sure if that's worth the churn though.

tgross35 replied:
Might as well since I'm updating those configs anyway :) Done in a fixup

RalfJung commented on library/coretests/tests/floats/f128.rs:
```diff
@@ -71,23 +60,23 @@ fn test_num_f128() {
     let ten = 10f128;
     let two = 2f128;
-    assert_eq!(ten.rem(two), ten % two);
+    assert_biteq!(ten.rem(two), ten % two);
```
Isn't `rem` the same as `%`? I'm a bit confused by the test.^^

Anyway I think this always worked in Miri ever nice basic arithmetic worked, there's no intrinsics involved here.

tgross35 replied:
This test came from https://github.com/rust-lang/rust/blob/8afd71079a729ac002298239f5010ccaac5807ca/library/coretests/tests/floats/mod.rs#L19-L35, indeed it is a bit dubious. I don't see anything else that exercises `Rem` but that would probably be something good to have.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15326628246/job/43122748605) [(plain)](https://github.com/rust-lang/rust/commit/0a879112914bd10be6d09df5d1eb63b0f664f9bd/checks/43122748605/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    floats::f128::test_recip

test result: FAILED. 1981 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 2.33s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.89.0-nightly\t(10a7d1b46\t2025-05-29)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:12:38
  local time: Thu May 29 15:16:19 UTC 2025
  network time: Thu, 29 May 2025 15:16:19 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented on library/coretests/tests/floats/f128.rs:
```diff
@@ -71,23 +60,23 @@ fn test_num_f128() {
 #[test]
 #[cfg(not(miri))]
 #[cfg(target_has_reliable_f128_math)]
 fn test_min_nan() {
```
Updated. powi, fma, and sqrt all still do not work, correct?

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15327547793/job/43125825121) [(plain)](https://github.com/rust-lang/rust/commit/04c4a4d18c8b87be717ea7bc0365f6173decabb3/checks/43125825121/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ureq v3.0.8
   Compiling citool v0.1.0 (/home/runner/work/rust/rust/src/ci/citool)
    Finished `dev` profile [unoptimized] target(s) in 7.45s
     Running `target/debug/citool calculate-job-matrix`
Run type: TryJob { job_patterns: Some(["aarch64-gnu", "gnu-aux"]) }
Error: Failed to calculate job matrix

Caused by:
    Patterns `gnu-aux` did not match any auto jobs
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented:
🤦 
@bors try

bors commented:
:hourglass: Trying commit 9c047e7b178ee46594206e040d964307648939b9 with merge a6090927d4f33b36e2998fdaa6d76908ba43c50e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9c047e7b178ee46594206e040d964307648939b9","merge_sha":"a6090927d4f33b36e2998fdaa6d76908ba43c50e"} -->

RalfJung commented on library/coretests/tests/floats/mod.rs:
```diff
@@ -15,6 +15,39 @@ macro_rules! assert_approx_eq {
+        let bits = (l.to_bits() - l.to_bits()).leading_zeros();
+        assert!(
+            l.to_bits() == r.to_bits(),
+            "{}\nl: {l:?} ({lb:#0width$x})\nr: {r:?} ({rb:#0width$x})",
```
This will print an empty line when the message is empty... do you have any good idea for avoiding that?

tgross35 replied:
I pushed a bit of a hacky way that adds a `\n` if the stringified tokens are nonempty. Open to better suggestions though, it would be nice if `format_args!` could be assigned to something.

tgross35 commented:
Thank you!
@bors r=RalfJung

bors commented:
:pushpin: Commit 70cce1c76264cfa7006d296a81ff2ae01afe186a has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 70cce1c76264cfa7006d296a81ff2ae01afe186a -->
<!-- homu: {"type":"Approved","sha":"70cce1c76264cfa7006d296a81ff2ae01afe186a","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
